ironfish/src/indexer/minedBlocksIndexer.ts

```diff
@@ -0,0 +1,268 @@
+      if (imported) {
+        this.chainProcessor.hash = null
+      }
+    })
+
+    this.accounts.onAccountRemoved.on(async (accountName) => {
+      await this.database.transaction(async (tx) => {
+        const hashes = await this.accountToBlockHashes.get(accountName, tx)
+        if (hashes) {
+          for (const hash of hashes) {
+            await this.minedBlocks.del(hash, tx)
+          }
+        }
+      })
+    })
+  }
+
+  async open(
+    options: { upgrade?: boolean; load?: boolean } = { upgrade: true, load: true },
+  ): Promise<void> {
+    if (this.isOpen) {
+      return
+    }
+
+    this.isOpen = true
+    await this.openDB(options)
+
+    if (options.load) {
+      await this.load()
+    }
+  }
+
+  async close(): Promise<void> {
+    if (!this.isOpen) {
+      return
+    }
+
+    this.isOpen = false
+    await this.closeDB()
+  }
+
+  async openDB(options: { upgrade?: boolean } = { upgrade: true }): Promise<void> {
+    await this.files.mkdir(this.location, { recursive: true })
+    await this.database.open()
+
+    if (options.upgrade) {
+      await this.database.upgrade(DATABASE_VERSION)
+    }
+  }
+
+  async closeDB(): Promise<void> {
+    await this.database.close()
+  }
+
+  async load(): Promise<void> {
+    const headHash = await this.meta.get('headHash')
+    this.chainProcessor.hash = headHash ? Buffer.from(headHash, 'hex') : null
+  }
+
+  start(): void {
+    if (this.isStarted) {
+      return
+    }
+    this.isStarted = true
+
+    void this.eventLoop()
+  }
+
+  async stop(): Promise<void> {
+    if (!this.isStarted) {
+      return
+    }
+    this.isStarted = false
+
+    if (this.eventLoopTimeout) {
+      clearTimeout(this.eventLoopTimeout)
+    }
+
+    if (this.database.isOpen) {
+      await this.updateHeadHash(this.chainProcessor.hash)
+    }
+  }
+
+  async eventLoop(): Promise<void> {
+    if (!this.isStarted) {
+      return
+    }
+
+    await this.updateHead()
+
+    if (this.isStarted) {
+      this.eventLoopTimeout = setTimeout(() => void this.eventLoop(), 1000)
+    }
+  }
+
+  async updateHead(): Promise<void> {
+    const { hashChanged } = await this.chainProcessor.update()
+
+    if (hashChanged) {
+      this.logger.debug(
+        `Updated MinedBlocksIndexer Head: ${String(this.chainProcessor.hash?.toString('hex'))}`,
+      )
+    }
+  }
+
+  async updateHeadHash(headHash: Buffer | null): Promise<void> {
+    const hashString = headHash && headHash.toString('hex')
+    await this.meta.put('headHash', hashString)
+  }
+
+  async getMinedBlocks(includeForks?: boolean): Promise<
+    {
+      main: boolean
+      hash: string
+      sequence: number
+      account: string
+      minersFee: number
+    }[]
+  > {
+    const minedBlocks = (await this.minedBlocks.getAllValues()).sort(
+      (a, b) => a.sequence - b.sequence,
+    )
+
+    if (includeForks) {
+      return minedBlocks
+    }
+
+    return minedBlocks.filter((block) => block.main === true)
+  }
```

Comment:
How do you filter on the sequence?

Also, is it possible for us to get away with just yielding blocks instead of accumulating and sorting all the blocks?

Reply:
My intention was to filter in the RPC route and have this strictly return the desired blocks, but there's really no reason that it couldn't be done here by passing in options instead. Regarding your second question, since we're using hashes as the key for the store, `.getAllValues()` ends up being lexicographically sorted by hash. The current output of `miners:mined` is sorted by sequence.